typescript/sdk/src/contracts.ts

```diff
@@ -1,79 +1,52 @@
-import { Contract } from 'ethers';
+import { ethers } from 'ethers';
 
-import { Router__factory } from '@abacus-network/app';
-import { AbacusConnectionManager__factory } from '@abacus-network/core';
 import { types } from '@abacus-network/utils';
 
-import { Connection, ProxiedAddress } from './types';
+import { Connection } from './types';
+import { objMap } from './utils';
 
-// address types generated from AbacusDeployer deployContract or deployProxiedContract
-export type AbacusContractAddresses = {
-  [key in string]: ProxiedAddress | types.Address;
+export type AbacusFactories = {
+  [key: string]: ethers.ContractFactory;
 };
 
-// from concrete ethers.ContractFactory static connect() type
-type EthersFactory<C> = (
-  address: types.Address, // TODO: generic on ProxiedAddress for proxy utilities
-  connection: Connection,
-) => C;
-export type Factories<A extends AbacusContractAddresses> = Record<
-  keyof A,
-  EthersFactory<any>
->;
+export type AbacusContracts = {
+  [key: Exclude<string, 'address'>]: ethers.Contract | AbacusContracts;
+};
 
-export interface IAbacusContracts<Addresses, Contracts> {
-  readonly addresses: Addresses;
-  readonly contracts: Contracts;
-  reconnect(connection: Connection): void;
-}
+export type AbacusAddresses = {
+  [key: string]: types.Address | AbacusAddresses;
+};
 
-export interface ContractsBuilder<
-  Addresses,
-  Contracts extends IAbacusContracts<Addresses, any>,
-> {
-  new (addresses: Addresses, connection: Connection): Contracts;
+export function addresses(contracts: AbacusContracts): AbacusAddresses {
+  return objMap(contracts, (_, contract): string | AbacusAddresses =>
+    'address' in contract ? contract.address : addresses(contract),
+  );
 }
 
-export abstract class AbacusContracts<
-  A extends AbacusContractAddresses,
-  F extends Factories<A> = Factories<A>,
-  CM = {
-    [key in keyof A]: F[key] extends EthersFactory<infer C> ? C : never;
-  },
-> implements IAbacusContracts<A, CM>
-{
-  abstract factories(): F;
-  // complexity here allows for subclasses to have strong typing on `this.contracts` inferred
-  // from the return types of the factory functions provided to the constructor
-  readonly contracts: CM;
-  constructor(readonly addresses: A, connection: Connection) {
-    const factories = this.factories();
-    const contractEntries = Object.entries(addresses).map(([key, addr]) => {
-      const contractAddress = typeof addr === 'string' ? addr : addr.proxy;
-      return [key, factories[key](contractAddress, connection)];
-    });
-    this.contracts = Object.fromEntries(contractEntries);
-  }
-
-  reconnect(connection: Connection): void {
-    Object.values(this.contracts).forEach((contract: Contract) =>
-      contract.connect(connection),
-    );
-  }
+export function attach(
```

Comment:
Similar like above. I get that you wanted to mimic the `attach` function on `ContractFactory` but maybe a more illustrative name would help as well?